README.md
# jQuery AutoCalc

Plugin to automate and simplify "on-the-fly" calculations.  The simplest use is to set the jsAutoCalc="" attribute on an input text tag and call `$("*").jAutoCalc();`

## Installation

`npm i -P jautocalc`

Includes:

- Development version with sourcemaps: 
	- `dist/jautocalc.js` - 24KB
	- `dist/jautocalc.js.map`
- Minimized production version: 
	- `dist/jautocalc.min.js` - 8KB

## Features

* The calculation is run when one of the value fields gains or loses focus or when the result field is changed.  Each calculation is also run once when the page is first loaded in case all the value fields come from the database.
* Field names are defined with braces, i.e. "{tax_amount}". If the field name is prefaced with "!" e.g. "{!tax_amount}" then the value will be used in the equation but it will ignore changes to that field for that equation only.
* Field names can be jquery selectors.
* Whitespace in the calculation is ignored.
* When multiple fields have the same name aggregate functions may be used.  The plugin supports:
	* SUM
	* AVG
	* MIN
	* MAX
	* COUNT - counts all fields
	* COUNTNOTEMPTY - counts all non-empty fields
* It can handle add, subtract, multiply, divide, and power (as **) in the proper order of operations.  It will respect parentheses.
* It can handle negative numbers and decimal numbers.
* If a calculation value field has no value yet, the result field is set to blank.
* when the result is calculated, the result field's change event is fired.  This is so calculations can be chained together.  In a shopping cart example, the grand total calculation relies on the subtotal calculation which relies on the sum of the line totals which rely on the quantity and amount fields.
* It will try to figure out what the currency symbol, thousands separator, and decimal separator are and, if found, reuse them in the result field.  By default, if it can't figure out the symbols it will use comma for thousands, period for decimal, and no currency symbol.

## Options

* attribute (string; default is `'jAutoCalc'`): The name of the attribute to look for to set up the auto calculation.
* thousandOpts (string array; default is `[',', '.', ' ']`): Possible options to use as the thousand separator.  Will use the first one it finds in the value fields to format the result field.  If you know this will always be the same, set the option to a single element array.
* decimalOpts (string array: default is `['.', ',']`): Possible options to use as the decimal separator.  Will use the first one it finds in the value fields to format the result field.  If you know this will always be the same, set the option to a single element array.
* decimalPlaces (number; default is `-1`): By default the plugin follows the rules for significant digits (the result is as significant as the least significant value).  There are cases when you might want to override this behavior, i.e. shopping cart where the item quantity will be whole numbers, but the item total should contain decimals.
* initFire (boolean; default is `true`): After configuring the value and results fields involved in a calculation, should the calculation be fired right away?  Useful if the value fields are pre-filled from some other source, e.g. database.
* chainFire (boolean: default is `true`): After calculating the result field value should the field be flagged as changed to potentially fire other calculations?
* keyEventsFire (boolean: default is `false`): Should the plugin do "instant-calculations" everytime keys are pressed in a value field?  By default, calculations are fired when focus is entered or lost on a value field.
* readOnlyResults (boolean: default is `true`): Should the plugin mark the result field(s) as read-only and be un-editable by the user?
* showParseError (boolean: default is `true`): Should the plugin show the parser errors as an alert box?  Useful for debugging/testing.
* emptyAsZero (boolean: default is `false`): empty values are treated as zero.
* smartIntegers (boolean: default is `false`): numbers like 123.000 treated as 123.
* onShowResult (function(el, value): default is `null`): called just before updating element with result.
* funcs (dictionary<string, IFunction>: default is `null`): user-defined functions.
* vars (dictionary<string, number>: default is `null`): user-defined constants.

## Working Example

```html
<html>
<head>
	<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js"></script>
	<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/jautocalc@1.3.1/dist/jautocalc.js"></script>
	<script type="text/javascript">
		<!--
		$(function() {

			function autoCalcSetup() {
				$('form#cart').jAutoCalc('destroy');
				$('form#cart tr.line_items').jAutoCalc({keyEventsFire: true, decimalPlaces: 2, emptyAsZero: true});
				$('form#cart').jAutoCalc({decimalPlaces: 2});
			}
			autoCalcSetup();


			$('button.row-remove').on("click", function(e) {
				e.preventDefault();

				var form = $(this).parents('form')
				$(this).parents('tr').remove();
				autoCalcSetup();

			});

			$('button.row-add').on("click", function(e) {
				e.preventDefault();

				var $table = $(this).parents('table');
				var $top = $table.find('tr.line_items').first();
				var $new = $top.clone(true);

				$new.jAutoCalc('destroy');
				$new.insertBefore($top);
				$new.find('input[type=text]').val('');
				autoCalcSetup();

			});

		});
		//-->
	</script>
</head>
<body>
<form id="cart">
	<table name="cart">
		<tr>
			<th></th>
			<th>Item</th>
			<th>Qty</th>
			<th>Price</th>
			<th>&nbsp;</th>
			<th>Item Total</th>
		</tr>
		<tr class="line_items">
			<td><button class="row-remove">Remove</button></td>
			<td>Stuff</td>
			<td><input type="text" name="qty" value="1"></td>
			<td><input type="text" name="price" value="9.99"></td>
			<td>&nbsp;</td>
			<td><input type="text" name="item_total" value="" jAutoCalc="{qty} * {price}"></td>
		</tr>
		<tr class="line_items">
			<td><button class="remove">Remove</button></td>
			<td>More Stuff</td>
			<td><input type="text" name="qty" value="2"></td>
			<td><input type="text" name="price" value="12.50"></td>
			<td>&nbsp;</td>
			<td><input type="text" name="item_total" value="" jAutoCalc="{qty} * {price}"></td>
		</tr>
		<tr class="line_items">
			<td><button class="remove">Remove</button></td>
			<td>And More Stuff</td>
			<td><input type="text" name="qty" value="3"></td>
			<td><input type="text" name="price" value="99.99"></td>
			<td>&nbsp;</td>
			<td><input type="text" name="item_total" value="" jAutoCalc="{qty} * {price}"></td>
		</tr>
		<tr>
			<td colspan="3">&nbsp;</td>
			<td>Subtotal</td>
			<td>&nbsp;</td>
			<td><input type="text" name="sub_total" value="" jAutoCalc="SUM({item_total})"></td>
		</tr>
		<tr>
			<td colspan="3">&nbsp;</td>
			<td>
				Tax:
				<select name="tax">
					<option value=".06">CT Tax</option>
					<option selected value=".00">Tax Free</option>
				</select>
			</td>
			<td>&nbsp;</td>
			<td><input type="text" name="tax_total" value="" jAutoCalc="{sub_total} * {tax}"></td>
		</tr>
		<tr>
			<td colspan="3">&nbsp;</td>
			<td>Total</td>
			<td>&nbsp;</td>
			<td><input type="text" name="grand_total" value="" jAutoCalc="{sub_total} + {tax_total}"></td>
		</tr>
		<tr>
			<td colspan="99"><button class="row-add">Add Row</button></td>
		</tr>
	</table>
</form>
</body>
</html>
```

See the working example at http://c17r.github.io/jAutoCalc/sample.html


CHANGELOG
1.4.1 - 31-Mar-2021
- update CHANGELOG and README

1.4.0 - 31-Mar-2021
- if a field name in an equation is prefaced with "!" e.g "{total} / {!qty}" the value of the field will still be used in the equation but a change to the field will not cause a recalculation.

1.3.2 - 31-Mar-2021
- Fix where not all jQuery events were namespaced properly

1.3.1 - 22-Mar-2021
- values are attempted to be parsed after cleansing to see if they are valid, stop process if the value is not valid.

1.2.1 - 17-Aug-2020
- If the calculation is fired but the new value is the same as the current, don't unnecessarily execute chainFire
- Fix IE11 compatibility

1.2.0 - 11-Jul-2018
- Broken up into separate files, webpack to package
- Converted from ES5 to TypeScript
- Prod config minifies
- Published on NPM

1.1.0 - 18-May-2013 (all from qyloxe)
- new function: countNotEmpty, counts number of non empty elements
- new option: emptyAsZero, empty input values are treated as zero
- new option: smartInteger, values 123.000 are converted to 123
- new event: onShowResult, fired just before value change
- new option: funcs, user defined functions
- new option: vars, user defined variables (Pi, etc)
- allows to select elements not only by name but by jQuery selector

1.0.0 - 20-Sep-2010
- initial release


dev-server.js
const express = require("express");
const { readFileSync } = require("fs");
const path = require("path");

const app = express();
const port = 8000;

const mix = (filename) => {
	const mixPath = path.resolve(__dirname, "./dist/mix-manifest.json");

    const entries = JSON.parse(readFileSync(mixPath, "utf-8"));
	const entry = entries[filename];
	if (!entry) {
		throw new Error(`${filename} not found in mix manifest.`);
	}

	const hotPath = path.resolve(__dirname, "./dist/hot");
	try {
		const hotUrl = readFileSync(hotPath, "utf-8");
		return `${hotUrl.trim()}${entry}`;
	} catch(err) {
		return entry;
	}
};

app
	.get('/', (req, res) => {
		res.send(`
<html>
    <head>
        <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
        <script type="text/javascript" src="${mix('/jautocalc.js')}"></script>
        <script type="text/javascript">
        <!--
            $(function() {

                function autoCalcSetup() {
                    $('form#cart').jAutoCalc('destroy');
                    $('form#cart tr.line_items').jAutoCalc({keyEventsFire: true, decimalPlaces: 2, emptyAsZero: true});
                    $('form#cart').jAutoCalc({decimalPlaces: 2});
                }
                autoCalcSetup();


                $('button.row-remove').on("click", function(e) {
                    e.preventDefault();

                    var form = $(this).parents('form')
                    $(this).parents('tr').remove();
                    autoCalcSetup();

                });

                $('button.row-add').on("click", function(e) {
                    e.preventDefault();

                    var $table = $(this).parents('table');
                    var $top = $table.find('tr.line_items').first();
                    var $new = $top.clone(true);

                    $new.jAutoCalc('destroy');
                    $new.insertBefore($top);
                    $new.find('input[type=text]').val('');
                    autoCalcSetup();

                });
                
                $('table#nonreactive tbody tr').jAutoCalc({
					initFire: false,
					readOnlyResults: false,
					chainFire: true,
					decimalPlaces: 2
                });

            });
        //-->
        </script>
    </head>
    <body>
        <form id="cart">
            <table name="cart">
                <tr>
                    <th></th>
                    <th>Item</th>
                    <th>Qty</th>
                    <th>Price</th>
                    <th>&nbsp;</th>
                    <th>Item Total</th>
                </tr>
                <tr class="line_items">
                    <td><button class="row-remove">Remove</button></td>
                    <td>Stuff</td>
                    <td><input type="text" name="qty" value="1"></td>
                    <td><input type="text" name="price" value="9.99"></td>
                    <td>&nbsp;</td>
                    <td><input type="text" name="item_total" value="" jAutoCalc="{qty} * {price}"></td>
                </tr>
                <tr class="line_items">
                    <td><button class="remove">Remove</button></td>
                    <td>More Stuff</td>
                    <td><input type="text" name="qty" value="2"></td>
                    <td><input type="text" name="price" value="12.50"></td>
                    <td>&nbsp;</td>
                    <td><input type="text" name="item_total" value="" jAutoCalc="{qty} * {price}"></td>
                </tr>
                <tr class="line_items">
                    <td><button class="remove">Remove</button></td>
                    <td>And More Stuff</td>
                    <td><input type="text" name="qty" value="3"></td>
                    <td><input type="text" name="price" value="99.99"></td>
                    <td>&nbsp;</td>
                    <td><input type="text" name="item_total" value="" jAutoCalc="{qty} * {price}"></td>
                </tr>
                <tr>
                    <td colspan="3">&nbsp;</td>
                    <td>Subtotal</td>
                    <td>&nbsp;</td>
                    <td><input type="text" name="sub_total" value="" jAutoCalc="SUM({item_total})"></td>
                </tr>
                <tr>
                    <td colspan="3">&nbsp;</td>
                    <td>
                        Tax:
                        <select name="tax">
                            <option value=".06">CT Tax</option>
                            <option selected value=".00">Tax Free</option>
                        </select>
                    </td>
                    <td>&nbsp;</td>
                    <td><input type="text" name="tax_total" value="" jAutoCalc="{sub_total} * {tax}"></td>
                </tr>
                <tr>
                    <td colspan="3">&nbsp;</td>
                    <td>Total</td>
                    <td>&nbsp;</td>
                    <td><input type="text" name="grand_total" value="" jAutoCalc="{sub_total} + {tax_total}"></td>
                </tr>
                <tr>
                    <td colspan="99"><button class="row-add">Add Row</button></td>
                </tr>
            </table>
        </form>
        <hr>
		<table id="nonreactive">
			<thead>
				<th>Quantity</th>
				<th>Price</th>
				<th>Total</th>
			</thead>
			<tbody>
				  <tr>
					<td><input name="qty" value="2" /></td>
					<td><input name="price" value="3" jAutoCalc="{total} / {!qty}" /></td>
					<td><input name="total" value="6" jAutoCalc="{qty} * {price}" /></td>
				  </tr>
			</tbody>
		</table>
    </body>
</html>

	`);
	})
	.use(express.static('dist'))
	;

app.listen(port, () => {
	console.log(`Listening on http://localhost:${port}`);
});
